plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,565 @@
+        }
+
+        return clientBuilder.build();
+    }
+
+    private static class S3ExchangeStorageWriter
+            implements ExchangeStorageWriter
+    {
+        private final S3Client s3Client;
+        private final S3AsyncClient s3AsyncClient;
+        private final String bucketName;
+        private final String key;
+        private final int partSize;
+        private final Optional<SecretKey> secretKey;
+
+        private int currentPartNumber;
+        private Optional<String> uploadId = Optional.empty();
+        private final List<CompletableFuture<CompletedPart>> uploadFutures = new ArrayList<>();
+
+        public S3ExchangeStorageWriter(S3Client s3Client, S3AsyncClient s3AsyncClient, String bucketName, String key, int partSize, Optional<SecretKey> secretKey)
+        {
+            this.s3Client = requireNonNull(s3Client, "s3Client is null");
+            this.s3AsyncClient = requireNonNull(s3AsyncClient, "s3AsyncClient is null");
+            this.bucketName = requireNonNull(bucketName, "bucketName is null");
+            this.key = requireNonNull(key, "key is null");
+            this.partSize = partSize;
+            this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        }
+
+        @Override
+        public ListenableFuture<Void> write(Slice slice)
+                throws IOException
+        {
+            // skip multipart upload if there would only be one part
+            if (slice.length() < partSize && uploadId.isEmpty()) {
+                PutObjectRequest.Builder putObjectRequestBuilder = PutObjectRequest.builder()
+                        .bucket(bucketName)
+                        .key(key);
+                S3RequestUtil.configureEncryption(secretKey, putObjectRequestBuilder);
+
+                try {
+                    s3Client.putObject(putObjectRequestBuilder.build(),
+                            // avoid extra memory copy
+                            fromContentProvider(() -> new ByteArrayInputStream(slice.getBytes()), slice.length(), Mimetype.MIMETYPE_OCTET_STREAM));
+                    return immediateVoidFuture();
+                }
+                catch (AwsServiceException e) {
+                    throw new IOException(e);
+                }
+            }
+
+            if (uploadId.isEmpty()) {
+                uploadId = Optional.of(createMultipartUpload().uploadId());
+            }
+
+            CompletableFuture<CompletedPart> uploadFuture = uploadPart(uploadId.get(), slice);
+            uploadFutures.add(uploadFuture);
+
+            return asVoid(toListenableFuture(uploadFuture));
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            if (uploadId.isEmpty()) {
+                return;
+            }
+
+            try {
+                List<CompletedPart> completedParts = uploadFutures.stream()
+                        .map(CompletableFuture::join)
```

Comment:
This will cause the `close` method to block for potentially a long time (as it will have to wait for all parts to finish upload). Currently if you look at the call stack it is not designed to block for a very long time. The call to `ExchangeSink#finish` is done from the `OutputBuffer#noMorePages` that is executed from a rather tiny executor that can easily run out of threads. 

This is actually my mistake. Now after carefully looking at the implementation of other output buffers it looks like `ExchangeSink#finish` and `ExchangeSink#abort` should be asynchronous and return a Future. I'm going to address this issue in the main PR.

Reply:
I will pay close attention to the changes in the main PR.